src/box/txn_limbo.c

```diff
@@ -649,10 +685,66 @@ txn_limbo_read_demote(struct txn_limbo *limbo, int64_t lsn)
 	return txn_limbo_read_promote(limbo, REPLICA_ID_NIL, lsn);
 }
 
+/**
+ * Check that some synchronous transactions have gathered quorum and
+ * write a confirmation entry of the last confirmed transaction.
+ */
+void
+txn_limbo_confirm(struct txn_limbo *limbo)
+{
+	if (limbo->is_in_rollback)
+		return;
+	struct txn_limbo_entry *e;
+	if (!limbo->confirmed_entry_is_valid) {
+		/* Try to find the next valid confirmed_entry value. */
+		e = rlist_next_entry(limbo->confirmed_entry, in_queue);
+		if (rlist_entry_is_head(e, limbo->queue, in_queue))
+			return;
+		for (; !rlist_entry_is_head(e, limbo->queue, in_queue);
+		       e = rlist_next_entry(e, in_queue))
+			if (txn_has_flag(e->txn, TXN_WAIT_ACK))
+				break;
+		limbo->confirmed_entry = rlist_prev_entry(e, in_queue);
+		/*
+		 * There is no next transaction or the next transaction has no
+		 * lsn. Therefore, confirmed_entry cannot be considered valid.
+		 */
+		if (rlist_entry_is_head(e, limbo->queue, in_queue) || e->lsn == -1)
+			return;
+		limbo->confirmed_entry_is_valid = true;
+		limbo->ack_count = vclock_count_ge(&limbo->vclock, e->lsn);
+	}
+	if (limbo->ack_count < replication_synchro_quorum)
+		return;
+	e = rlist_next_entry(limbo->confirmed_entry, in_queue);
+	/* Make sure confirmed_entry is valid. It must be true. */
+	assert(!rlist_entry_is_head(e, limbo->queue, in_queue));
+	assert(txn_has_flag(e->txn, TXN_WAIT_ACK));
+	assert(e->lsn != -1);
```

Comment:
The limbo can have txns not having an LSN yet and also being synchronous. It happens when their WAL write isn't finished yet, but they already received ACKs and a confirm. How does this assert survive that?

Reply:
This is the invariant we provide.
`confirmed_entry_is_valid == true` only when the next transaction after `confirmed_entry` is synchronous and has lsn (not -1) and `ack_count` stores the correct value.
The `confirmed_entry_is_valid` variable is written to `true` in a single place in the code and only if lsn != -1. https://github.com/Astronomax/tarantool/blob/6eee079a537a40c9721f6ca8a9fe616748f3b5c6/src/box/txn_limbo.c#L689
If at any point this is no longer true, `confirmed_entry_id_valid` is immediately set to `false`. This is processed in all places where the limbo is modified in any way.

Reply:
Right, thanks for the explanation.